dll/win32/shell32/wine/shellpath.c

```diff
@@ -208,29 +215,61 @@ PathQualifyExW(_Inout_ LPWSTR pszPath, _Inout_opt_ LPCWSTR pszDir, _In_ DWORD dw
                 ++pchTemp;
         }
 
-#if 0 // FIXME: IsLFNDriveW has a bug
         bLFN = IsLFNDriveW(szRoot);
-#endif
         if (!bLFN)
         {
-            /* FIXME: Support non-LFN */
+            PWCHAR pch;
+            for (pch = pchTemp; *pch != UNICODE_NULL; ++pch)
+            {
+#define PATH_CHAR_CLASS_DOT         0x00000004
+#define PATH_CHAR_CLASS_BACKSLASH   0x00000008
+#define PATH_CHAR_CLASS_COLON       0x00000010
+#define PATH_CHAR_CLASS_OTHER_VALID 0x00000100
+#define VALID_SHORT_PATH_CHAR_CLASSES ( \
+    PATH_CHAR_CLASS_DOT | \
+    PATH_CHAR_CLASS_BACKSLASH | \
+    PATH_CHAR_CLASS_COLON | \
+    PATH_CHAR_CLASS_OTHER_VALID \
+)
+                if (!PathIsValidCharW(*pch, VALID_SHORT_PATH_CHAR_CLASSES))
+                {
+                    *pch = L'_';
+                }
+            }
         }
 
         StringCchCopyW(pszPath, MAX_PATH, szRoot);
         cchPathLeft -= lstrlenW(pszPath) + 1;
     }
     else /* UNC path: Begins with double backslash */
     {
-#if 0 // FIXME: IsLFNDriveW has a bug
         bLFN = IsLFNDriveW(pchTemp);
-#endif
-        if (!bLFN)
+        if (bLFN)
+        {
+            pszPath[2] = UNICODE_NULL; /* Cut off */
+            cchPathLeft -= (2 + 1);
+            pchTemp += 2;
+        }
+        else
         {
-            /* FIXME: Support non-LFN */
+            PWCHAR pchSlash = StrChrW(&pszPath[2], L'\\');
```

Comment:
```suggestion
            PWCHAR pchSlash = StrChrW(pszPath + 2, L'\\');
```
Is done a line below as well, and is more readable imo.

Reply:
OK.